sycl/doc/extensions/experimental/sycl_ext_oneapi_bindless_images/ZE_experimental_bindless_images.asciidoc

```diff
@@ -0,0 +1,621 @@
+
+== Terminology
+
+For the purposes of this document, a bindless image is one which provides
+stateless access to the underlying data via image reference handles. At the
+application level, this allows the user to implement programs where the number
+of images is not known at compile-time, and store all handles to images --
+irrespective of varying formats and layouts -- in some container, e.g. a dynamic
+array.
+
+== Overview
+
+In this document, we propose the following changes to the Level Zero API:
+
+- The creation of images to be split into explicit allocation of image memory
+  and the creation of image handles from the previously allocated memory
+- Support for creation of images on linearly allocated memory backed by USM
+- Addition of image memory copy functions that allow copying to and from image
+  memory before an image handle's construction
+
+Currently, in Level Zero, the `zeImageCreate` performs the image memory
+allocation and image handle generation. Moreover, this function only allows for
+the allocation of image memory in an implementation-specific layout. Delivering
+the features proposed in this document will improve the functionality and
+flexibility of the Level Zero API.
+
+
+=== Background
+
+The DPC++ bindless images extension has sought to provide the flexibility of
+bindless images at the SYCL application level. Per the proposal, users would be
+able to separate image memory allocation from the image handle creation.
+Tools such as SYCLomatic would benefit from this flexibility when converting
+CUDA to SYCL code. Level Zero's support for this extension would enable
+converted code to run optimally on Intel hardware.
+
+Additionally, per the DPC++ bindless images extension, users would be able to
+create and sample images based on linearly allocated memory. A valuable use-case
+is the creation of a bindless image on already existing USM memory, preventing
+copies, and hardware sampling can be applied to that data.
+
+This proposal aims to extend Level Zero to better align with the PI API used by
+the SYCL runtime in DPC++.
+
+
+=== Specification
+
+To enable the separation of image memory allocation from image handle
+generation, we propose two new allocation functions `zeMemAllocImageExp` and
+`zeMemAllocPitchedExp`.
+
+`zeMemAllocImageExp` will allocate non-USM image memory with an implementation
+specific layout (such as tile swizzle patterns or lossless compression).
+
+Whereas, `zeMemAllocPitchedExp` will allocate USM memory in a linear layout.
+Memory allocated with this API should be freed using `zeMemFree`.
+
+Alongside these allocation functions, we propose a singular additional freeing
+function, `zeMemFreeImageExp`, to free memory allocated with
+`zeMemAllocImageExp`.
+
+We also propose a new struct `ze_device_image_bindless_exp_desc_t` that
+describes device restrictions on image properties. It should be passed to
+`ze_device_properties_t::pNext`. It contains the fields `imagePitchAlign`,
```

Comment:
are the new images limited to 2D images only ?

or are 3D images allowed too ?

for 3D images - slice pitch has to be defined ( size of 2D image + padding )

Reply:
Yes, we do want to support 3D USM images. If this requires alignment restrictions on both rowPitch and slicePitch, then this will need to be added. This may include adding any or all of the following:
- imageSlicePitchAlign
- maxImageLinearDepth
- maxImageLinearSlicePitch

I'm not sure if these names are the best choice, but the idea stands.